sdk/src/trace/batch_span_processor.cc

```diff
@@ -157,8 +158,30 @@ void BatchSpanProcessor::Export(const bool was_force_flush_called)
                     });
                   });
 
-  exporter_->Export(nostd::span<std::unique_ptr<Recordable>>(spans_arr.data(), spans_arr.size()));
+  /* Call the sync Export when force flush was called, even if
+     is_export_async_ is true.
+  */
+  if (is_export_async_ == false)
+  {
+    exporter_->Export(nostd::span<std::unique_ptr<Recordable>>(spans_arr.data(), spans_arr.size()));
 
+    NotifyForceFlushCompletion(was_force_flush_called);
+  }
+  else
+  {
+    exporter_->Export(nostd::span<std::unique_ptr<Recordable>>(spans_arr.data(), spans_arr.size()),
+                      [this, was_force_flush_called](sdk::common::ExportResult result) {
```

Comment:
Should we declare async `Export` of `LogExporter` and `SpanExporter` as `void Export(const nostd::span<std::unique_ptr<Recordable>> &records, std::function<bool(ExportResult)>&& result_callback) noexcept override` ?
These code capture some data and this lambda object can not be destroyed before `result_callback` be called. We need store this `result_callback` some where and call it later, but with `function_ref`, we can not move it.

@lalitb @DebajitDas 

> As metioned in http://open-std.org/JTC1/SC22/WG21/docs/papers/2019/p0792r5.html , `function_ref` is designed as a parameter, but not for storage.

Reply:
yes good point. we can't transfer ownership of `function_ref`.
`std::function<bool(ExportResult)>&& result_callback` would be right callback to use.

Reply:
@DebajitDas Could you please raise another PR to modify the type of `result_callback` and push to `async-changes`? I will rebase from it merge the async version of `OtlpHttpClient::Export` then.

Reply:
Sure @owent, will raise by EOD

Reply:
Raised this PR https://github.com/open-telemetry/opentelemetry-cpp/pull/1278 which has the changes for type change

Reply:
> Raised this PR #1278 which has the changes for type change

Thanks I have raise #1281 to implement async `OtlpHttpExporter` and `OtlpHttpLogExporter`. Both async and sync unit tests are also added for these two exporters.